src/VASSAL/counters/Translate.java

```diff
@@ -151,43 +152,151 @@ public String myGetType() {
 
   @Override
   public Command keyEvent(KeyStroke stroke) {
-    myGetKeyCommands();
-    if (moveCommand.matches(stroke)) {
-      // Delay the execution of the inner piece's key event until this piece has moved
-      return myKeyEvent(stroke);
-    }
-    else {
-      return super.keyEvent(stroke);
+    // Classic MFD delays the execution of the inner piece's key event until after this piece has moved
+    // This unexpectedly changes the order of trait execution, but is required for the old Move Batcher to work correctly
+    if (GlobalOptions.getInstance().isUseClassicMoveFixedDistance()) {
+      myGetKeyCommands();
+      if (moveCommand.matches(stroke)) {        
+        return myKeyEvent(stroke);
+      }
     }
+    // For New MFD, use standard trait execution timing
+    return super.keyEvent(stroke);
   }
 
   @Override
   public Command myKeyEvent(KeyStroke stroke) {
     myGetKeyCommands();
-    Command c = new NullCommand();
     if (moveCommand.matches(stroke)) {      
-      if (mover == null) {
-        mover = new MoveExecuter();
-        mover.setKeyEvent(stroke);
-        mover.setAdditionalCommand (setOldProperties());
-        SwingUtilities.invokeLater(mover);
-      }
-      GamePiece target = findTarget(stroke);
-      if (target != null) {
-        c = c.append(moveTarget(target));
-      }
-      mover.addKeyEventTarget(piece);
-      // Return a non-null command to indicate that a change actually happened
-      // Note: Looks weird to wipe out the Commands, but they have all been added to the Move Executor.
-      c = new NullCommand() {
-        @Override
-        public boolean isNull() {
-          return false;
+      return GlobalOptions.getInstance().isUseClassicMoveFixedDistance() ? classicTranslate(stroke) : newTranslate(stroke);
+    }
+    return null;
+  }
+  
+  /*
+   * New Translate code.
+   * Simplified. Get rid of Move Batcher. Use same technique as Send To Location to
+   * move counters. A series of Translate commands issues by a Trigger Action will now act as
+   * expected and Undo properly.
+   * 
+   * NOTE: If the Stack Move option is used and a series of MFD commands are issued by a Trigger Action
+   * then the moving pieces will 'pick up' any pieces they land on along the way. The Stack Move option is not
+   * recommended for this reason and now defaults to 'N' and is marked in the Editor as 'Not Recommended'.
+   */
+  protected Command newTranslate (KeyStroke stroke) {
+    Command c = new NullCommand();
+    
```

Comment:
There's no point in creating the `Command` this early when you could return without ever using it. You could create it on line 217, for example.

Reply:
True. Change incorporated.